### DISCUSSION TITLE
ahooks 使用过程中，针对 eslint 需要做什么处理，才能让 useEffect 的依赖列表不用加入相关的依赖？

### DISCUSSION THREAD

**Discussion by ghost** [alibaba/hooks#2252] [Q&A] [answered]
https://ahooks.js.org/zh-CN/guide/blog/function#ahooks-%E5%87%BD%E6%95%B0%E5%A4%84%E7%90%86%E8%A7%84%E8%8C%83

![image](https://github.com/alibaba/hooks/assets/125900423/41f004f1-94ac-49ae-abb1-9876ca3fdb08)

在 ahook 官网给出的 ahooks **函数处理规范** 中提到：

> ahooks 所有 Hooks 返回的函数，都拥有和 setState 一样的特性，地址不会变化，放心大胆的使用即可。

但是如果在 useEffect 中使用函数，会收到 eslint 规则给出提示：

> React Hook useEffect has a missing dependency: 'setData1'. Either include it or remove the dependency array.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)

代码示例：
```js

  const [data1, setData1, getData1] = useGetState(1)
  useEffect(() => {
    setData1(2)
  }, [])

  const [data2, setData2] = useState(1)
  useEffect(() => {
    setData2(2)
  }, [])

```

![image](https://github.com/alibaba/hooks/assets/125900423/01e1f90d-6f40-4061-a303-8f3398fbdde1)

![image](https://github.com/alibaba/hooks/assets/125900423/6060ae6a-288f-43ef-96b8-42ac2611b95f)

所以针对 eslint 需要做什么处理，才能让 useEffect 的依赖列表不用加入相关的依赖？
（前提是不写 eslint-disabled）

**Comment by liuyib** [comment, marked as answer]
目前没有办法区分，这是 react 相关生态的问题。只能通过把 error 改成 warn 来缓解：

```js
{
  "plugins": [
    // ...
    "react-hooks"
  ],
  "rules": {
    // 加上这个
    "react-hooks/exhaustive-deps": "warn"
  }
}
```

不想这样干的话，那就按照这个插件提示的那样，把需要放入依赖数组的东西都放进去。

---

依赖数组里的参数应该放多少，这是个颇具有争议的事情，[useEffectEvent](https://react.dev/reference/react/experimental_useEffectEvent) 这个新 hook 的出现（不需要写依赖数组），貌似就是来解决这个争议的，见：[https://zhuanlan.zhihu.com/p/514751791](https://zhuanlan.zhihu.com/p/514751791)（注：文中的 useEvent 就是这个 [useEffectEvent](https://react.dev/reference/react/experimental_useEffectEvent)，更换了名字，见：https://github.com/facebook/react/pull/25881）。

**Reply by ghost** [reply]
翻了一下 react 仓库中 [eslint-plugin-react-hooks]-exhaustive-deps] 的代码，

在 [isStableKnownHookValue](https://github.com/facebook/react/blob/fdc8c81e07e9b3b49604afe524ee6d130fd3e03e/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L164) 方法中对放入依赖数组中的内容作出了判断说明

![image](https://github.com/alibaba/hooks/assets/125900423/1cfd3dfa-3dfa-4820-80b3-bab6befcad0a)

简单说，除了几个固定的规则外，其他的都认为是 dynamic 动态的值。动态的值没有放入依赖数组就会报错

针对 ahooks 这种**明知不会变化的**值是不做判断的。

社区有一些相关的提案，期望用户能自行声明哪些值是可以不放入数组的 static 值，但官方似乎不愿意接受。

有第三方的fork修改方案，又或者可以自己修改代码
https://github.com/facebook/react/issues/16873#issuecomment-536346885


----

另外：关于依赖数组的争议，可以看官方的说明
https://github.com/facebook/react/issues/14920#issuecomment-471070149

![image](https://github.com/alibaba/hooks/assets/125900423/f9a7aaca-4aa0-4cbe-87e9-a73925601a58)

相关链接：

官方对 exhaustive-deps 的说明
[[ESLint] Feedback for 'exhaustive-deps' lint rule](https://github.com/facebook/react/issues/14920#top)

自定义依赖数组提案相关：
https://github.com/facebook/react/issues/24112
https://github.com/facebook/react/issues/16873
https://github.com/facebook/react/issues/19330

**Reply by liuyib** [reply]
> 社区有一些相关的提案，期望用户能自行声明哪些值是可以不放入数组的 static 值，但官方似乎不愿意接受。

这个方案不错，目前官方不支持没办法。个人认为 useEffectEvent 是解决依赖数组问题的好东西。只能等待官方和社区一点点优化开发体验吧~

**Reply by ghost** [reply]
useEffectEvent (又或者说是原来的 useEvent），目前社区提案中实现方式和 ahooks 的 useMemorizedFn 类似，

也许现在可以用 useMemorizedFn 作为代餐

**Reply by liuyib** [reply]
useMemoizedFn 在未来的版本就是等价于 useEffectEvent，只不过目前没做这个处理（有计划的）。所以现在使用 useMemoizedFn 相当于可以在 <18 版本的 react 中用到 useEffectEvent 的能力。

不过 useMemoizedFn 能解决多少问题，就仁者见仁吧，依赖数组问题现状不太明朗，期待官方彻底优化吧~

**Reply by liuyib** [reply]
从这里 https://github.com/facebook/react/issues/14920#issuecomment-1117600014 可以看出，useEffectEvent (旧：useEvent) 就是 react 官方为了解决依赖数组问题提出的新 hook。

mark 一下
